src/monitor.ts

```diff
@@ -0,0 +1,425 @@
+    // data must be a JSON object with the following format.
+    // {
+    //   nodes: [IMonitoredNodeConfig...]
+    // }
+    this.app.post('/add', (req, res) => {
+
+      winston.info(req.body);
+
+      if (!req.body.hasOwnProperty('nodes')) {
+        res.status(400).send('There is no field named nodes.');
+        return;
+      }
+
+      if (!(req.body.nodes instanceof Array)) {
+        res.status(400).send('The field nodes is not an array.');
+        return;
+      }
+
+      const validConfigs = [];
+      try {
+        for (const data of req.body.nodes) {
+          const monitoredNodeConfig = validateMonitoredNodeConfig(data);
+          this.nodesStatus.push(this.configToInitialNodeStatus(monitoredNodeConfig));
+          validConfigs.push(monitoredNodeConfig);
+          this.appendConfigData(validConfigs);
+          res.status(200).send('Ok');
+        }
+      } catch (err) {
+        res.status(400).send(`${req.body} contains invalid configuration, error: ${err}`);
+      }
+    });
+  }
+
+  private tryRevive(node: IMonitoredNodeStatus) {
+    if (node.failureCount < this.failureTolerance) {
+      winston.info(`${node.address} hasn't reached failure threshold yet, not reviving...`);
+      return;
+    }
+
+    if (typeof node.reviveCmd === 'undefined') {
+      winston.info(`${node.address} doesn't have a revive command...`);
+      return;
+    }
+
+    let args: string[] = [];
+    if (typeof node.reviveArgs !== 'undefined') {
+      args = node.reviveArgs;
+    }
+
+    const child = ChildProcess.spawn(node.reviveCmd, args, {
+      detached: true,
+      stdio: 'ignore',
+    });
+
+    child.unref();
+  }
+
+  private async ping() {
+    for (const node of this.nodesStatus) {
+      // https://github.com/ethereum/wiki/wiki/JSON-RPC#net_version
+      // We use this simple API as a PING to the network.
+
+      try {
+        const body = await rp.post(`http://${node.address}`, {
+          headers: {
+            'content-type': 'application/json',
+          },
+          json: {
+            jsonrpc: '2.0',
+            method: 'net_version',
+            params: [],
+            // We should ignore any response whose id is smaller than the last
+            // response id to linearize the communication between the monitor
+            // and the network nodes.
+            id: node.lastResponseId + 1,
+          },
+        });
+
+        const nodeResp = this.validateNetVersionResponse(body);
+        if (nodeResp.id > node.lastResponseId) {
+          // Ignore stale responses by checking the request/response id
+          node.alive = true;
+          node.networkId = nodeResp.result;
+          node.lastUpdate = new Date();
+          node.lastResponseId = nodeResp.id;
+        }
+      } catch (err) {
+        winston.error(`Failed to contact node at ${node.address}, reason: ${err}`);
+        node.alive = false;
+        node.failureCount++;
```

Comment:
is this reset somewhere?

Reply:
Now it's reset everytime node.alive becomes true